src/processor.cpp

```diff
@@ -0,0 +1,401 @@
+    //TODO: Skill used
+
+    void load( JsonObject &jo, const std::string &src );
+    void check() const;
+
+};
+
+process_id get_process_id( std::string id )
+{
+    return process_id( id );
+}
+
+generic_factory<process_data> processes_data( "process type", "id" );
+
+}
+
+void process_data::load( JsonObject &jo, const std::string & )
+{
+    mandatory( jo, was_loaded, "name", name );
+    //TODO: modify to allow lists of components
+    mandatory( jo, was_loaded, "components", components,  string_id_reader<itype_id> {});
+    mandatory( jo, was_loaded, "output", output );
+    mandatory( jo, was_loaded, "fuel_intake", fuel_intake );
+    mandatory( jo, was_loaded, "duration", duration );
+}
+
+void process_data::check() const
+{
+    const process_id proc_id( id.str() );
+
+    if( !proc_id.is_valid() ) {
+        debugmsg( "There is no process with the name %s", id.c_str() );
+    if( !components.is_valid() ) {
+        debugmsg( "Invalid components \"%s\" in \"%s\".", components.c_str(), id.c_str() );
+    }
+    if( !output.is_valid() ) {
+        debugmsg( "Invalid output \"%s\" in \"%s\".", output.c_str(), id.c_str() );
+    }
+    if( !fuel_intake.is_valid() ) {
+        debugmsg( "Invalid fuel intake \"%s\" in \"%s\".", fuel_intake.c_str(), id.c_str() );
+    }
+    if( !duration.is_valid() ) {
+        debugmsg( "Invalid duration \"%s\" in \"%s\".", duration.c_str(), id.c_str() );
+    }
+}
+
+void process_data::load( JsonObject &jo, const std::string &src )
+{
+    processes_data.load( jo, src );
+}
+
+void process_data::check()
+{
+    processes_data.check();
+}
+
+void process_data::reset()
+{
+    processes_data.reset();
+}
+
+void interact_with_processor( const tripoint &examp, player &p )
+{
+    //Get processor id at examp
+    processor_id pid = get_processor_id( examp );
+    //Do I need the debugmsg-es here or do they get checked in process_data::load?
+    if( !processors_data.is_valid( pid ) ) {
+        debugmsg( "Examined %s has action interact_with_processor, but has no processor data associated", g->m.furn( examp ).id().c_str() );
+        return;
+    }
+    //Get data associated with said processor id
+    const processor_data &processor = processors_data.obj( pid );
+    std::vector<auto> possible_processes;
+    for( auto &p : processor.processes() ) {
+        if( p.is_valid() ){
+            possible_processes.push_back( process_data.obj( process_id( p ) ) );
+        } else {
+            //Debugmsg probably not exactly right,
+            //I'm trying to return the processor name and the problematic process name too
+            debugmsg( "Examined %s has non-valid process data %s2 associated", processor.id().c_str(), get_process_id( p ).c_str() );
+            return;
+        }
+    }
+
+    //  Figure out which process to run, easier to work with than
+    //  trying to figure out desired process from items on tile,
+    //  Can become a mess with processor with multiple possible processes
```

Comment:
True. I think I figured out a way to do this.

- If there's only one process, use that.
- If there are items, check each process component see whether it's a subset of the items present, and take the process with the largest matching subset as active. Currently only works with exact subsets of items, as in if we have item A,B and C present and need A and B for the process, then it picks that, unless there's a process utilizing all 3 as input. I need to implement it in a way that plays nice with interchangeable components / materials. (think kiln possibly using a mixture of bones and wood)
- If there are no items, ask the player what they want to do and then look in their inventory for components

Method two should work for figuring out which process we are in the middle of, for the .is_processing method